In [0]:
!python --version

Python 3.6.9


[slide](https://drive.google.com/drive/folders/1cAM251bjoBCYF2bHfMM07MOGEgU2Q2VQ?usp=sharing)

[gitHub](https://github.com/Pierian-Data/Complete-Python-3-Bootcamp)

[Source](https://www.udemy.com/course/complete-python-bootcamp)


[Reading](https://www.python-course.eu/python3_re.php)






* integers: whole number

1 2 3 

* floats: numbers with decimal point

1.2 3.2 2.4

* String: ordered immutable sequence of characters (immutable:you can't reassign characters)

"a banana" 'an orange'

* Tuples: (immutable lists) ordered immutable sequence of objects. Tuples are similar to list but they are immutable. Once an element is inside a tuple it can't be changed/reassigned.

(10,"hello",200.3)

* lists: Ordered mutable sequence of objects. Objects can be rereived by location. ordered sequence can be indexed or sliced.

[10,"hello",200.3]

* Dictionaries:  unordered key,value pairs. Dictionaries are unordered and can't be sorted. they provide mappings betwen k and v for storing objects. The key-value pair allows users to quickly grab objects without needing to know the index location.

{"key":"value", "key2":"value2"}

* sets: Unorderd mutable collection of  unique elements (hashable objects)

    * set contents need to be immutable to be hashable. So you can have a set of
tuples, but not a set of lists.

{"10","hello","200.3"}
{(1,2),(3,4),(5,6)}

* boolean: logical value indicating True False

True, False




objects are iterable : we can iterate over every element in object
string and list are iterable, we can iterate over key indictionary


In [0]:
'''
Itterable
'''
my_iterable_string='123'
my_iterable_list=[1,2,3]
my_iterable_tuple=(1,2,3)
dictt={'hh':1,'gg':11}
my_iterable_dict=dictt.items() #.keys() .values()
index=0
x=my_iterable_dict
for item_name in x:
    
    print(index,item_name)
    index += 1



0 ('hh', 1)
1 ('gg', 11)


In [0]:
stringg='abc123abc'
stringg='11'
listt = [1,2,3]
tuplee = (1,2,3)
sett= {1,2,(1,3)}
dictt = {'1':1,'2':2,('1','3'):3}
dictt


{('1', '3'): 3, '1': 1, '2': 2}

In [0]:
stringg='ab\tcd caef\nch'

#adjust by replacing certain value with another in all string or join parts
a='..'.join(stringg) #'a..b..\t..c..d.. ..c..a..e..f'
a.replace('..','--') #'a--b--\t--c--d-- --c--a--e--f'

#adjust by split or patition
stringg.partition('c') #('ab\t', 'c', 'd caef')
stringg.split('ca') #['ab\tcd ', 'ef']
stringg.split('c') #splits on the string or char: ['ab\t', 'd ', 'aef']
stringg.split() #splits on space and tab : ['ab', 'cd', 'caef']
stringg.splitlines() #['ab\tcd caef', 'ch ']


# count and find/index/endswith patterns
stringg.count('a') #2
stringg.endswith('bca') #True
stringg.startswith('ca') #False
stringg.find('a') #return index 0
stringg.find('ca',4,8) # returns 6 ('string,,start,endindex)
stringg.index('ca',4,8) # returns 6 ('string,,start,endindex)
# The index() method is similar to find() method for strings. 
# The only difference is that find() method returns -1 if the substring is not found, whereas index() throws an exception.
stringg.find('z',2) #if not found return -1
stringg.rfind('z',2) #if not found return -1
#stringg.index('z') # if not found gives error****Important****


#adjust the look by adjusting appearance to right or left
stringg.zfill(10) #zero fill on the left:'0000000abc'
stringg.expandtabs(10) #replacetabs with tabsize=N spaces
stringg.ljust(17,'*') #(widthLen, char)'ab\tcd caef\nch ***'
stringg.rjust(17,'*') #'****ab\tcd caef\nch'

#adjust by removing space
stringg.lstrip() #remove white space on left side
stringg.strip()
stringg.rstrip()

#adjust by playing with caps case
stringg.lower()
stringg.title()
stringg.swapcase()
stringg.upper()

# check status of the Caps case and/or contentent
# stringg.isalnum()
# stringg.isalpha()
# stringg.isdecimal()
# stringg.islower()
# stringg.isnumeric()
# stringg.isspace()
# stringg.istitle()
# stringg.isupper()

# convert string to list
list(stringg) #['a', 'b', '\t', 'c', 'd', ' ', 'c', 'a', 'e', 'f', '\n', 'c', 'h']

#string in any of these followoing is** commands produce false except for alphanymeric

# here a,b and c are characters and 1,2 and 3 
# are numbers 
stringgg= "abc123"
print("abc123: ", stringgg.isalnum()) 
  
# here a,b and c are characters and 1,2 and 3  
# are numbers but space is not a alphanumeric  
# character 
stringgg = "abc 123" 
print("abc 123: " ,stringgg.isalnum())

'stringg'.isalpha() #true
'string '.isalpha() #false
'string1g '.isalpha() #false cuz of space

'111'.isalnum() #true
'11c1'.isalnum() #true
'11c1 '.isalnum() #false
' '.isalnum() #false
'11.1'.isalnum() #false

'111'.isdecimal() #true
'11.01'.isdecimal() #false

'111'.isnumeric() #true
'11.1'.isnumeric() #false

a='bn'
b=a
a=a.ljust(17,'*')
a == b  #false

# 'Oranges  |    10     |        10'
'{0:<8} | {1:^9} | {1:>9}'.format('Oranges', 10, 'bananas\n', 'Oranges', 10, 'bananas\n')

# 'Oranges | 10 | bananas\n'

'{} | {} | {}'.format('Oranges', 10, 'bananas\n', 'Oranges', 10, 'bananas\n') 


abc123:  True
abc 123:  False


'Oranges  |    10     |        10'

In [0]:
c=[5,3,32,1,55] 

# only for lists reverse and sort
# c.reverse # inplace reverse
# c.sort # inplace sort

# c.index and .count for all itterables string/list/tuple
# c.count #same as string and list and tuple
# c.index #c.index for all iterables (value,start,end) same as list, string, tuples (string has find too)


# list append/extend/insert corresponds to add in set
# c.append # appends to the end as it is
# c.extend # appends to the end only itterable elementrs one char at a time
# c.insert # insert object before index : (index,object)

# c.pop # for mutables only: inplace c.pop(indexx) similar to list [] dict {} nand set {}
# c.remove # for removing an element from the list[] and set{} c.remove(oneChar) remove only one char by value in place
# a=c.copy  # **list same as **dictionary and **set


print(sorted(c))
print(reversed(c))

c.sort() # in place [1, 3, 5, 32, 55]
print(c)

c.reverse() #in place [55, 32, 5, 3, 1]
print(c)

c.append(2)
c.append('gf') #[55, 32, 5, 3, 1, 2, 'gf']
c.append((1,2)) #[55, 32, 5, 3, 1, 2, 'gf',(1,2)]
c.append([1,2]) #[55, 32, 5, 3, 1, 2, 'gf',[1,2]]
c.append({'1':2}) #[55, 32, 5, 3, 1, 2, 'gf',{'1':2}]
c.append(1)
c=c[:4]
print(c)

c.extend('gf') #[55, 32, 5, 3, 1, 2, 'gf', 'g', 'f']
c.extend([1,2]) #[55, 32, 5, 3, 1, 2, 'gf', 'g', 'f', 1, 2]
print(c)
c.insert(0,'gf')

print(c)

#remove only one char by indexx in place
indexx=3
c.pop(indexx)
c.pop(-1)
print(c)

#remove only one char by value in place
c.remove('g')
print('removeee',c)

#c.sort() Error sort not supported between instances of 'int' and 'str'
#print(c) 

c.count(2)
print(c.count(2))

b=c.copy()
c.append('h') #creates copy
print(len(b)==len(c))
b=c #creates pointer
c.append('gh')
print(len(b)==len(c))

c.pop(2)
print(c)



[1, 3, 5, 32, 55]
[1, 3, 5, 32, 55]
[55, 32, 5, 3, 1]
[55, 32, 5, 3]
[55, 32, 5, 3, 'g', 'f', 1, 2]
['gf', 55, 32, 5, 3, 'g', 'f', 1, 2]
['gf', 55, 32, 3, 'g', 'f', 1]
removeee ['gf', 55, 32, 3, 'f', 1]
0
False
True
['gf', 55, 3, 'f', 1, 'h', 'gh']


In [0]:
# tuple()
#.index and .count for all itterabkes string/list/tuple
# c.index
# c.count

c=(1,2,3,1,2,1,3)
c.count(1) #2
# c.index(val,includedstart,excludedend) #returns first index with value
c.index(1) #0
c.index(1) #0
c.index(1,1,-1) # 3 ***** Error if x not in tuple
c.index(1,3) # 3
c.index(1,4) # 5
c.index(3,5) #6
# c.index(3,5,-1) #Get an error/value at last element so by excluding the end of the string by putting -1 


6

In [0]:
dictt = {'1':1,'2':2,'3':3}
print(dictt)
dictt.pop('2') # dictt.pop() gives error
print(dictt)
dictt.popitem()
print(dictt)
dictt = {'1':1,'2':2,'3':3}
print(dictt.values())
print(dictt.keys())
print(dictt.items())


{'1': 1, '2': 2, '3': 3}
{'1': 1, '3': 3}
{'1': 1}
dict_values([1, 2, 3])
dict_keys(['1', '2', '3'])
dict_items([('1', 1), ('2', 2), ('3', 3)])


In [0]:
sett={1,2,3}

# set([1,3])   | set([4,6])    #{1,3,4,6}
# set([1,4,3]) & set([4,6])    #{4}
# set([1,4,3]) - set([4,6])    #{1, 3}
# set([4,6])   - set([1,4,3])  #6

# sett.pop
# sett.remove
# sett.copy
# sett.add

# sett.intersection
# sett.union
# sett.difference
# sett.discard
# sett.symmetric_difference

print(sett)
sett.pop() # no index or value available
print(sett)
sett.remove(2) # remove value available
print(sett)
sett.add(2)
print(sett)
sett.add(2)
print(sett)
sett.add(22)
print(sett)
a=b=sett.union(sett)
print(a)
sett.difference(sett)
print(b)
c=sett.intersection(sett)
print(c)
print(sett.symmetric_difference(sett))

x = {"apple", "banana", "cherry"}
y = {"google", "microsoft", "apple"}

print(x.intersection(y))
print(x.union(y))
print(x.difference(y))
print(x.symmetric_difference(y))

# sett.difference
# sett.union
# sett.symmetric_difference
# sett.intersection
# sett.issubset
# sett.issuperset
# sett.isdisjoint
# sett.pop
# sett.remove

{1, 2, 3}
{2, 3}
{3}
{2, 3}
{2, 3}
{2, 3, 22}
{2, 3, 22}
{2, 3, 22}
{2, 3, 22}
set()
{'apple'}
{'microsoft', 'google', 'banana', 'apple', 'cherry'}
{'banana', 'cherry'}
{'google', 'banana', 'cherry', 'microsoft'}


In [0]:
'b'=='B' #false
2.0 == 2 # true
2 != 3 #true
1 < 2 #true
1<=2 #true
1>=2 #false

2 != 3 and 2 != 3 
2 != 3 or 2 != 3

1<2>3 #false
1<2 and 2> 3 #false
(1<2 ) and (2> 3) #false

not 1==1
not(1==1)



bmn


In [0]:
# only two
min(1,2)

# as many in the list
min([1,2,3])

1

In [0]:

hungry=True
if hungry:
    a=input()
    print(a,"bmn")
elif 1==1:
    print("jl")
else:
    print('n,m')



kj
kj bmn


In [0]:
def makes_twenty(n,m):
    return (n+m)==20 or n==20 or m==20

n,m=20,0
makes_twenty(n,m)

True

In [0]:
def myfunc(st):
    print(type(st))
    a=st[0].upper() + st[1:2] +st[3].upper()+st[3:]
    return a
myfunc('bananna')


<class 'str'>


'BaAanna'

In [0]:
def textrevereser(text):
    a=text.split()
    a.reverse()
    return str(a)
print(textrevereser('I am home already'))

['already', 'home', 'am', 'I']


## Extending the string


In [0]:
def string3x(st):
    '''
    Given a string return a string that every char repeated 3X
    '''
    st3=''
    for i in st:
        st3+=3*i
    return st3
print(string3x("Happy Monday!"))

HHHaaappppppyyy   MMMooonnndddaaayyy!!!


In [0]:
a,b,c=1,2,3
sum([a,b,c])
sum([a,b])

3

In [0]:
'''
return sum of array element except ignore sections between 6 and 9
'''

t=0
a=[0, 3, 6, 9]
b=a.copy()
for i in a:
    print(i,a,b)
    if i==6 or t==1:    
        t=1
        b.remove(i)
        if i==9:
            t=0
            continue
   
    
print(f"sum of {b} is {sum(b)}")

0 [0, 3, 6, 9] [0, 3, 6, 9]
3 [0, 3, 6, 9] [0, 3, 6, 9]
6 [0, 3, 6, 9] [0, 3, 6, 9]
9 [0, 3, 6, 9] [0, 3, 9]
sum of [0, 3] is 3


In [0]:
def myfunc(a):
    b=''
    for i in a:
        if i==0 or i==7:
            b+=str(i)
            if '007' in b:
                return True
    return False

a=[1,2,0,4,0,8,7,9]
myfunc(a)

True

In [0]:
#learn if its float or not

a=9/(9**0.5)
int(3.0)==3

i=9
list(range(2,int(i**0.5)+1))



SyntaxError: ignored

In [0]:
print ("Displaying list of numbers by reverse order using range()")
list(range(4,-1,-1)) # [4, 3, 2, 1, 0]
list(range(16,12,-1)) # [16, 15, 14, 13]
list(range(4,0,-1)) # [4, 3, 2, 1]

Displaying list of numbers by reverse order using range()


[4, 3, 2, 1]

In [0]:
def myfunc(a=[17,4,2,3,1,6,16]):
    '''
    Prime number: can be devided by only itself and one
    Prime NO. Z= X*Y X=1 and Y=Z
    loop through from 1:sqrt(z) and see if z/(1:sqrt(z))
    '''

    prime=[]
    for i in a:
        for j in range(int(i**0.5),0,-1):
            # print(f"testing {i}, with {j}")
            # figures out if number is devidable by other numbers
            if j==1: #23=23*1
                    prime.append(i)
                    #print(f"{i} is prime")       
            elif int(i/j) == i/j: #20/4=5 5==5.0
                break
    return prime


            
print(f"List of all primes: {myfunc()}")
            


List of all primes: [17, 2, 3, 1]


Floating-point numbers are represented in computer hardware as base 2 (binary) fractions. 
https://docs.python.org/2/tutorial/floatingpoint.html

C language converts floating point numbers to fractions and when those fractions are called by Python they again are converted to float and the conversion between a float and fraction are not always accurate. Round() will round the floating point number to a certain number to the right of the decimal which we can control in the syntax.

In [0]:
round(0.1 + 0.2 - 0.3,1)   

0.0

In [0]:
print("I am a\nbanana")

I am a
banana


In [0]:
0.1 + 0.2 - 0.3

5.551115123125783e-17

**Python Operators: Order & Precedence**
In order of highest to lowest Precedence
* President	()	Parentheses
* Powers	**	Exponentiation
* Asterisk	*	Multiplication
* Slash	/	Division
* Plus	+	Addition
* Minus	–	Subtraction

In [0]:
mystring="Heyo pretty world!"
print(mystring[-2])
print(mystring[:-2])
print(mystring[1:3])
print(mystring[2:])
print(mystring[1:5:2])
print("backward reverse:",mystring[::-1])
print(mystring[::2])
print(mystring[:])
print(mystring[::])
print(mystring[3::])
print("Hello\tWorld"[7])
print("Hello World"[7])
print("Hello\nWorld"[7])


d
Heyo pretty worl
ey
yo pretty world!
eo
backward reverse: !dlrow ytterp oyeH
Hy rtywrd
Heyo pretty world!
Heyo pretty world!
o pretty world!
o
o
o



String Formatting
String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation. As a quick comparison, consider:

In [0]:
# immutable means can't change
name = "SAm"
# name[0]="p"
# you get an error cause strings are immutable
newname='P'+'Sam'[1:]
print(newname)

newname='P'*3+'Sam'[1:]
print(newname)

Pam
PPPam


In [0]:
a= 'string '*3
print(a.title())
print(a.upper())
print(a.split())
print(a.split('s'))


String String String 
STRING STRING STRING 
['string', 'string', 'string']
['', 'tring ', 'tring ', 'tring ']


In [0]:
print('This is a string {}'.format('blaaaa'))

print("The ordered {} {} {}".format('bla','blaa','blaaa'))

print("The reverse {2} {1} {0}".format('bla','blaa','blaaa'))

print("The reverse {c} {a} {b}".format(a='bla',b='blaa',c='blaaa'))

print("The reverse {c} {c} {c}".format(a='bla',b='blaa',c='blaaa'))


This is a string blaaaa
The ordered bla blaa blaaa
The reverse blaaa blaa bla
The reverse blaaa bla blaa
The reverse blaaa blaaa blaaa


In [0]:
result=104.12345
print(result)
print("the result {r}".format(r=result))
print("the result {}".format(result))
print("float formatting rule: {value:width.precisionf}")

print("the result {r:1.3f}".format(r=result))
print("the result {r:10.3f}".format(r=result))
print("the result {r:10.5f}".format(r=result))


104.12345
the result 104.12345
the result 104.12345
float formatting rule: {value:width.precision f}
the result 104.123
the result    104.123
the result  104.12345
his name is Jose


In [0]:

name = "Jose"
age = 3
print("his name is {} and he is {} years old".format(name,age))

print(f"his name is {name} and he is {age} years old")

his name is Jose and he is 3 years old
his name is Jose and he is 3 years old


Alignment, padding and precision with .format()
Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more


By default, .format() aligns text to the left, numbers to the right. You can pass an optional <,^, or > to set a left, center or right alignment:


Also you can precede the aligment operator with a padding character

In [0]:
print('{0:8} | {1:9}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:9}'.format('Apples', 3.))
print('{0:8} | {1:9}'.format('Oranges', 10))
print('{0:8} | {1:9}'.format('Oranges', 10))
print("the result {rez:10.5f}".format(rez=result))


print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))



print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))


num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))


Fruit    | Quantity 
Apples   |       3.0
Oranges  |        10
Oranges  |        10
the result  104.12345
Left     |  Center  |    Right
11       |    22    |       33
Left==== | -Center- | ...Right
11====== | ---22--- | ......33
My 10 character, four decimal number is:   23.4500


In [0]:
num=3.565
name='hgjfhgj'
print(f"My 10 character, four decimal number is:{num:10.4}")
print(f'helloo   \t {name:<8} | {name:^8} | {name:>8}  ')
print(f'helloo   \t {name:.<8} | {name:.^8} | {name:.>8}  ')

print(f"{'mon':<8} \t {'tue':^8} \t {'wed':>8}")

a,b,c,d,e,f,g="mon","tue","wed","thu","fri","sat","sun"
print(f"{a:<8} \t {b:^8} \t {c:>8}")
print(f"{d:<8} \t {e:^8} \t {f:>8}")



My 10 character, four decimal number is:     3.565
helloo   	 hgjfhgj  | hgjfhgj  |  hgjfhgj  
helloo   	 hgjfhgj. | hgjfhgj. | .hgjfhgj  
mon      	   tue    	      wed
mon      	   tue    	      wed
thu      	   fri    	      sat


In [0]:
'{} {}'.format('one', 'two')



In [0]:
'{:f}'.format(3.141592653589793)

'3.141593'

In [0]:
'{:0.2f}'.format(33.0798798)

'33.08'

In [0]:
'{:d}'.format(33)

'33'

Named placeholders
Both formatting styles support named placeholders.

In [0]:
data = {'first': 'Hodor', 'last': 'Hodor!'}
'{first} {last}'.format(**data)



'Hodor Hodor!'

In [0]:
'{:{align}{width}}'.format('test', align='^', width='10')
f"{'test':<8}"


'hjh     '

In [0]:
aa='test'
f"{aa :<8}"

'test    '

In [0]:
data = [4, 8, 15, 16, 23, 42]
'{d[4]} {d[5]}'.format(d=data)

In [0]:
print(f"{'python rules' :8>} | {'python rules' :8>}")


python rules | python rules


In [0]:
f"{'python rules' :8>} |  {'python rules' :8>}"

'python rules |  python rules'

In [0]:
f"{'bla' :5} | {'bli':5}"

'bla   | bli  '

In [0]:
f"{'blii' :.<5} | {'blu' :.<5}"

'blii. | blu..'

In [0]:
f"{'a':.<6} | {'b':.^7} | {'c':.>6}"

'a..... | ...b... | .....c'

List is a mutable ordered list of objects

In [0]:
a=['one',2,3]

b=['onee',22,33]
len(a)
a.append('appended item')
c=a+b
print("before mutation:",c)
c[3]='four'
print("after mutation:",c)

d=c.pop()
print("popped item:",d)
print("newlist:",c)

d=c.pop(1)
print("popped item in index 1:",d)
print("newlist:",c)

c=[5,3,32,1,55]

print("sorted in place the following list",c)
print('does not change and modify the list after showing its sorted version',sorted(c))
print('list stays the same unsorted',c)
c.sort()
print('final list after inplace sorting: ', c)

print("reversed in place the following list",c)
c.reverse()
print('final list: ', c)



before mutation: ['one', 2, 3, 'appended item', 'onee', 22, 33]
after mutation: ['one', 2, 3, 'four', 'onee', 22, 33]
popped item: 33
newlist: ['one', 2, 3, 'four', 'onee', 22]
popped item in index 1: 2
newlist: ['one', 3, 'four', 'onee', 22]
sorted in place the following list [5, 3, 32, 1, 55]
does not change and modify the list after showing its sorted version [1, 3, 5, 32, 55]
list stays the same unsorted [5, 3, 32, 1, 55]
final list after inplace sorting:  [1, 3, 5, 32, 55]
reversed in place the following list [1, 3, 5, 32, 55]
final list:  [55, 32, 5, 3, 1]


In [0]:
a=['one',2,3]
c='abc'
print("another way to reverse the list like string")
print(f"reversed list: {a[::-1]} and reversed string: {c[::-1]}")


another way to reverse the list like string
reversed list: [3, 2, 'one'] and reversed string: cba


access nested list element: 1. How do I index a nested list? For example if I want to grab 2 from [1,1,[1,2]]?

You would just add another set of brackets for indexing the nested list, for example: my_list[2][1] . We'll discover later on more nested objects and you will be quizzed on them later!

In [0]:
a=[1,2,[3,4]]
a[2][0]

3

In [0]:
my_dict = {'k1':'v1','k2':'v2','k3':[10,12,22,'apple'], 'k4':{'insideKey':2}}
my_dict['k1'] #v1
my_dict['k3'][0] #10
my_dict['k3'][3].upper() #'Apple'
my_dict['k4']['insideKey'] #2

shopping_receipt={'apples':1.99, 'banana':2.99}
print(f"banana price is: {shopping_receipt['banana']}")

shopping_receipt['pineapples']=3
shopping_receipt['pineapples']=3.50
print(shopping_receipt)

shopping_receipt.items() #tuples (key,value)
shopping_receipt.keys()
shopping_receipt.values()
shopping_receipt.pop('banana','notfound')
print(shopping_receipt)
shopping_receipt.popitem()

print(shopping_receipt.get('apple','not fount'))
print(shopping_receipt.get('apples','not fount'))
print(shopping_receipt['apple'])
print(shopping_receipt)



banana price is: 2.99
{'apples': 1.99, 'banana': 2.99, 'pineapples': 3.5}
{'apples': 1.99, 'pineapples': 3.5}
not fount
1.99


KeyError: ignored

Tupples are ordered sequence of items

In [0]:
mytuple=(1,2,3,1,11,12)
mylist=[1,2,3,1,11,12]

len(mytuple)
mytuple[0]
mytuple.count(1)
firstTimeIndex = mytuple.index(1)
reversedtuple= mytuple[::-1]
print(reversedtuple)
print(mytuple[:-5:-1])
print(mytuple[2:5:2])
print(mytuple[-2])




(12, 11, 1, 3, 2, 1)
(12, 11, 1, 3)
(3, 11)
11


In [0]:
myemptytuple2= ()
print(type(myemptytuple2))
myemptytuple= tuple()
print(type(myemptytuple))

<class 'tuple'>
<class 'tuple'>


sets are unordered collection of unique elements


In [0]:
myset = set()
myset.add(1)
myset.add(11)
myset.add(2)
myset.add(22)
print(myset)
myset.remove(22)
print(myset)
mysett=myset.copy()
myset.add(22)
print(myset)
print(mysett)
listt = [11,11,11,2,3,33,33]
list(set(listt)) #remove repetition


{1, 2, 11, 22}
{1, 2, 11}
{1, 2, 11, 22}
{1, 2, 11}


[3, 33, 2, 11]

In [0]:
type(False)

bool

In [0]:
1>2

False

In [0]:
1==1

True

In [0]:
b = None # itis a placeholder


In [0]:
%%writefile myfile.txt
Hellloooo
whatuppp 
banaaaaana applesss orangesss


Writing myfile.txt



* mode='r'   read
* mode='r+'  read+write
* mode='w'   write ( if file exists: overwrites existing file; if not exist creates a new file)
* mode='w+'  read+write ( if file exists: overwrites existing file; if not exist creates a new file)
* mode='a'   only appends (will add on to file)  



In [0]:
myfile = open('myfile.txt')
print(myfile.read()) #read(): string is output
# to read again you need to reset the curser back to the begining of file
myfile.seek(0)
print(myfile.readlines()) # readlines() : list is output
myfile.seek(0) 
print(myfile.readline())  #readline(): string is output
myfile.close()

# to avoid closing an openning we do the following

with open('myfile.txt') as myfile:
    contents= myfile.read()
    print(contents)

with open('myfile.txt','r') as myfile:
    print( myfile.read())


with open('myfile.txt','a') as myfile:
    myfile.write("\n added line")

with open('myfile.txt','r') as myfile:
    print(myfile.read())

with open('myfile2.txt','w') as myfile2:
    myfile2.write('bla\nblaa\nblaa')

with open('myfile2.txt','r') as myfile2:
    print(myfile2.read())


Hellloooo
whatuppp 
banaaaaana applesss orangesssadded line
 added line
 added line
 added line
 added line
 added line
['Hellloooo\n', 'whatuppp \n', 'banaaaaana applesss orangesssadded line\n', ' added line\n', ' added line\n', ' added line\n', ' added line\n', ' added line']
Hellloooo

Hellloooo
whatuppp 
banaaaaana applesss orangesssadded line
 added line
 added line
 added line
 added line
 added line
Hellloooo
whatuppp 
banaaaaana applesss orangesssadded line
 added line
 added line
 added line
 added line
 added line
Hellloooo
whatuppp 
banaaaaana applesss orangesssadded line
 added line
 added line
 added line
 added line
 added line
 added line
bla
blaa
blaa


In [0]:
with open('myfile3.txt','w+') as myfile3:
    myfile3.write('blaaaa\nblaaaaaaaa\n')
    myfile3.seek(0)  # Important: return to the top of the file before reading,
    print(myfile3.read())

with open('myfile3.txt','w+') as myfile3:
    myfile3.write('blaaaa\nblaaaaaaaa\n')
    myfile3.seek(0)  # Important: return to the top of the file before reading,
    print(myfile3.read())

with open('myfile3.txt','r+') as myfile3:
    myfile3.write('blaaaa\nblaaaaaaaa\n')
    myfile3.seek(0)  # Important: return to the top of the file before reading,
    print(myfile3.read())
   
with open('myfile3.txt','a') as myfile3:
    myfile3.write('blaaaa2\nblaaaaaaaa2\n')

with open('myfile3.txt','r') as myfile3:
    print(myfile3.read())

blaaaa
blaaaaaaaa

blaaaa
blaaaaaaaa

blaaaa
blaaaaaaaa

blaaaa
blaaaaaaaa
blaaaa2
blaaaaaaaa2



# Generators vs Functions

* range(start,end,stepsize) 

* zip(list1,list2)

* enumerate(list)

* range is a generator is a type of function that generates the values instead of saving it in to the memory

* hence its more efficient


In [0]:
mylist=[4,5,6,7,8]
three=len(mylist)
startindex=2
endindex=three+2        
for num in range(0,8,2):
    print(num)
print('--'*5)
for i in range(startindex,endindex,2):
    print(i)
print('--'*5)
for i in range(len(mylist)):
    print(i,mylist[i])
print('--'*5)
for i in mylist:
    print(i)

list(range(2,11,2)) # [2, 3, 4, 6, 8, 10]

0
2
4
6
----------
2
4
6
----------
0 4
1 5
2 6
3 7
4 8
----------
4
5
6
7
8


[2, 4, 6, 8, 10]

In [0]:
index_count=0
for letter in 'abcd':
    print(f"at {index_count} the letter is {letter}")
    index_count+=1
# at 0 the letter is a
# at 1 the letter is b
# at 2 the letter is c
# at 3 the letter is d
for item in enumerate('abcd'):
    print(item)
# (0, 'a')
# (1, 'b')
# (2, 'c')
# (3, 'd')
for index,letter in enumerate('abcd'):
    # print(index)
    # print(letter)
    print(index,letter)
    # 0 a
    # 1 b
    # 2 c
    # 3 d
for index,letter in enumerate([8,9,5]):
    # print(index)
    # print(letter)
    print(index,letter)
    # 0 8
    # 1 9
    # 2 5

at 0 the letter is a
at 1 the letter is b
at 2 the letter is c
at 3 the letter is d
(0, 'a')
(1, 'b')
(2, 'c')
(3, 'd')
0 a
1 b
2 c
3 d
0 8
1 9
2 5


In [0]:
list1=list(range(3))
list2=['a','b','c']
list3=list(range(5))

#generators
print("list of a range(start,end,step)):",list(range(1,10,2))) #[1, 3, 5, 7, 9]
print("list of enumerate one list",list(enumerate(list1))) #[(0, 0), (1, 1), (2, 2)]
print("list of enumerate one list",list(enumerate(list1,2))) #[(2, 0), (3, 1), (4, 2)]
print("list of zipped three lists",list(zip(list1,list2,list3))) # [(0, 'a', 0), (1, 'b', 1), (2, 'c', 2)]

for a,b,c in zip(list1,list2,list3):
    print(a,b,c)



list of a range(start,end,step)): [1, 3, 5, 7, 9]
list of enumerate one list [(0, 0), (1, 1), (2, 2)]
list of enumerate one list [(2, 0), (3, 1), (4, 2)]
list of zipped three lists [(0, 'a', 0), (1, 'b', 1), (2, 'c', 2)]
0 a 0
1 b 1
2 c 2


# item in itterable


In [0]:
print('a' in 'abc')
print('a' in {'a':1})
print('a' in {'a','b'})
print(1 in {'a':1}.values())
print('a' in ['a','b','c'])


True
True
True
True
True


#min and max and rand

In [0]:
mylist=[1,2,5,7]
min(mylist)
max(mylist)

7

In [0]:
from random import shuffle, randint

mylist=[1,2,5,7,8,6,7,3,0]
print("before shuffle:",mylist)

shuffle(mylist) # in place shuffling
print("after shuffle:",mylist)

mynum=randint(3,10) #grab rand number in range of [a,b]
print(mynum)

result=input('Hi User: Enter a float or an int number: ')
result=float(result)+3.4 
print(result)
result=input('Hi User: Enter only an int number: ')
result=int(result)+3 #
print(result)

before shuffle: [1, 2, 5, 7, 8, 6, 7, 3, 0]
after shuffle: [7, 6, 3, 0, 8, 2, 5, 1, 7]
3
Hi User: Enter a float or an int number: 24
Hi User: Enter only an int number: 7
10


## List comprehension


In [0]:
[i for i in 'apple' if i!='A' and i!='p'] #['a', 'l', 'e']
[num**2 for num in range(1,10,2) if num%2==0] #[1, 9, 25, 49, 81]

celcius = [0,10,20, 34, 100]
fahrenheit = [((9/5) * temp + 31) for temp in celcius] #[31.0, 49.0, 67.0, 92.2, 211.0]
print(fahrenheit)

#[0, '0DD', 2, '0DD', 4, '0DD', 6, '0DD', 8, '0DD', 10]
[x if x%2 == 0 else '0DD' for x in range(0,11)] #only if else goes at beggining



[31.0, 49.0, 67.0, 92.2, 211.0]


[0, '0DD', 2, '0DD', 4, '0DD', 6, '0DD', 8, '0DD', 10]

# methods and builtin objects

* To get documentation on object you 

help(mylist.insert)

* Mylist.insert+press (shift+tab ) in jupyter notebook

In [0]:
def my_function():
    '''
    DESCRIPTION: Info about my func
    DATE WRITTEN: 8th of FEB 2020
    INPUT: name
    OUTPUT: Hello name
    '''
    name_user=input("Pleae enter your name: ")
    return(name_user.title())

print(f"Function description: {help(my_function)}")
print(f"Hello {my_function()}")



Help on function my_function in module __main__:

my_function()
    DESCRIPTION: Info about my func
    DATE WRITTEN: 8th of FEB 2020
    INPUT: name
    OUTPUT: Hello name

Function description: None
Pleae enter your name: m gh hj
Hello M Gh Hj


In [0]:
def say_hello(name='Name'):
    print(f'Hello {name}')

say_hello() #Hello Name
say_hello('Bella') #Hello Bella

def say_hello(name='Name'):
    return f'Hello {name}'
print(f"{say_hello('Lucas')}")



Hello Name
Hello Bella
Hello Lucas


In [0]:
# def dog_finder(st=''):

#     '''
#     DESCRIPTION: findout if the word "dog" is in the string?
#     INPUT: string
#     OUTPUT: TRUE/FALSE
#     '''

#     if len(st)==0:
#         return False

#     else:
#         for i in st.split():
#             if i.lower()=="dog":
#                 return True

def dog_finder(st=''):

    '''
    DESCRIPTION: findout if the word "dog" is in the string?
    INPUT: string
    OUTPUT: TRUE/FALSE
    '''
    return "dog" in st
              

n=input("please type: ")
print(f"It is {dog_finder(st=n)} that there is a dog in the sentense")

please type: hj gh dog
It is True that there is a dog in the sentense


In [0]:
def myfunc(a,b):
    '''
    returns 5% of the sum of a and b
    a,b are two positioal arguments 
    '''
    return sum((a,b))*0.05

# help(myfunc)
myfunc(40,60)

#what if we want to work with more than two numbers??
# maybe user puts more than two arguments
#when a function parameter starts with {*} 
#we treat it as a tuple of arbitrary number arguments

Help on function my_fun in module __main__:

my_fun(a, b)
    returns 5% of the sum of a and b



# \*args and **kwargs

## passing arbitrary number of arguments or keyworded arguments: with

* \*args = arguments
* **kwargs = Keyword arguments

When you want to accept an arbitarary number of arguments and keyword arguments without having to pre-define a bunch of parameters in your function calls


In [0]:
def myfunc(*args):
    #args returns back a TUPLES
    print(args)
    for item in args:
        print(item)

myfunc(4,5,6,7,8)

(4, 5, 6, 7, 8)
4
5
6
7
8


In [0]:
def myfunc(**kwargs):
    #kwargs returns a dictionary
    print(kwargs)
    if 'fruit' in kwargs:
        print(f"my fruit of choice is {kwargs['fruit']}")
    else:
        print('I did not find any fruit name')

myfunc(fruit='apple',veggie='lettuce')


{'fruit': 'apple', 'veggie': 'lettuce'}
my fruit of choice is apple


{6}

In [0]:
def myfunc(*args,**kwargs):
    print(args)
    print(kwargs)
    print(f"I would like {args[0]} {kwargs['food']}")
#myfunc(position is important args shoud be first kwargs should be second to match function above
myfunc(10,20,30, fruits='orange', food='eggs', animal='dogs')

(10, 20, 30)
{'fruits': 'orange', 'food': 'eggs', 'animal': 'dogs'}
I would like 10 eggs


# Lambda expression, Map and Filter

* Map function: you pass along the function you want to apply to every element in the itterable (e.g list) and avoid the for loop. (Map is generator so we need list to see the contrents) 
    * map(func,*iterableObject): map applies the func to every eement in the iterator
    * To use: map(funcNameWithoutParanthesis,ListName)
    * To see generator output using list use: list(map(myfunc,mylist))


* Filter function: returns an iterator yielding those items of the iterable than when passed through the function the result is True. You need to use a function that returns true or false
    * filter(function or None, iterable)
    * see generator output using list: list(filter(funcWithBoolOutput, iterable))
    * filter is going to filter based on the functions condition


* Lamda expression (anonomous function):



In [0]:
#square func for [1,2,3,4,5]
def square(num):
    return num**2
for i in [1,2,3,4,5]:
    print(square(i))

#replace for loop with Map
list(map(square,[1,2,3,4,5]))

1
4
9
16
25


[1, 4, 9, 16, 25]

In [0]:
def splicer(mystring):
    if len(mystring)%2 == 0:
        return 'Even'
    else:
        return mystring[0]

names=['Andy','Eve','Sally']

# use splicer than splicer() so map executes it not you

list(map(splicer,names)) #['Even', 'E', 'S']

#replace myfunc with map and lambda to only get first letter for not even length names
mylist=['Andy','Eve','Sally']
list(map(lambda names: names[0], mylist)) #['A', 'E', 'S']

#replace myfunc with map and lambda to only get reverse of names
list(map(lambda i: i[::-1] , mylist)) #['ydnA', 'evE', 'yllaS']


['ydnA', 'evE', 'yllaS']

In [0]:
#Boolean function return true or false
def check_even(num):
    if num%2==0:
        return num

#Instead use the filter function
# when you want to filter out of 
# a contion is true for ypur inputs

#Boolean function return true or false
def check_even(num):
    return num%2==0
mynums = [1,2,3,4,5]

#it filters based on the function and returns items that are result in True
list(filter(check_even,mynums)) # output [2,4]

#replace myfunc with lambda and apply filter to it
list(filter(lambda num: num%2==0, mynums))  # output [2,4]

[2, 4]

In [0]:
#step 1
def square(num):
    result = num ** 2
    return result

#step 2
def square(num):
    return num ** 2

#step 3
def square(num): return num ** 2

#step 4
def square(num): return num ** 2

#step 5
lambda num: num ** 2

#step 6 (optional)
square = lambda num: num ** 2

square(5)

#step 7 (usualy used lambda with Map and filter)
#map(func,mylist)
#lambda funcInput: funcReturnedOutput
mynums=[1,2,3,4,5]
list(map(lambda num:num**2,mynums))

[1, 4, 9, 16, 25]

#usualy used lambda with Map and filter

* map(func,mylist)
* lambda funcInput: funcReturnedOutput
* combined: map(lambda funcInput: funcReturnedOutput, mylist)

#LEGB Rule for variable scope assignment (Local Enclosing Global Built-in based on variable scope)


In Python, the LEGB rule is used to decide the order in which the namespaces are to be searched for scope resolution.
The scopes are listed below in terms of hierarchy(highest to lowest/narrowest to broadest):

* Local(L): Defined inside function/class
    * (e.g name in this) 
```
    
    def greet():    
        def hello():
            name ='Local string' #Local
            print('Hello I am a ',name)
        hello()
 ```
```

* Enclosed(E): Defined inside enclosing functions(Nested function concept)
    * (e.g name in this) 
    ```
    def greet():
        name ='I am Enclosing string' #Enclosing
        def hello():
            print('Hello',name)
        hello()
    ```

* Global(G): Defined at the uppermost level
    * (e.g name in this) 
```
    name ='I am a Global string' #Global
    def greet():    
        def hello():
            print('Hello',name)
        hello()
 ```
* Built-in(B): Reserved names in Python builtin modules
    

In [0]:
x=50

def funcc():
    print(f'X is {x}')
    #local reassignment on global variable
    x="New Value"
    print(f'I just  locally changed global X to {x}')

def func():
    global x
    print(f'X is {x}')
    #local reassignment on global variable
    x="New Value"
    print(f'I just  locally changed global X to {x}')

print(x) #50
print(func()) # X is 50 // I just  locally changed global X to New Value
print(x) #New value
#print(funcc()) #UnboundLocalError: local variable 'x' referenced before assignment


50
X is 50
I just  locally changed global X to New Value
None
New Value


#Avoid using global re-assignment!
Use the following return variable instead of using global re-assignment

In [0]:
x=50
def func():
    global x
    print(f'X is {x}')
    #local reassignment on global variable
    x="New Value"
    print(f'I just  locally changed global X to {x}')
    return x

x=funx() 

In [0]:
'''
Compute sphere volume:
Solve for volume
V=(4/3)*π*r^3
'''
r=3
volume=(4/3)*3.14*r**3
f"{volume:0.2f}" 


r=3
volume = lambda r : (4/3)*3.14*r**3
print(f"{volume(r):0.2f}")


113.04


In [0]:
def inlist(i,mylist=[3,7,9,4,6]):
    return i in mylist


i=[4,5,6]
v=list(filter(inlist,i))
print(v)


[4, 6]


In [0]:
'''
Count Upper Case and Lower Case strings
'''
string='ORANGE or APPLE or BANNANA'

mydict=dict()

keys=set(str)

for i in string:
    if i not in mydict.keys():
        mydict[i]=1
    else:
        mydict[i]+=1

mydict



{' ': 4,
 'A': 5,
 'B': 1,
 'E': 2,
 'G': 1,
 'L': 1,
 'N': 4,
 'O': 1,
 'P': 2,
 'R': 1,
 'o': 2,
 'r': 2}

In [0]:
'''
No. of upper and lower case chars
'''
string='ORANGE or APPLE or BANNANA'
d={'upper':0,'lower':0}
for c in string:
    if c.isupper():
        d['upper']+=1
    if c.islower():
        d['lower']+=1
    else:
        pass

print(f"org string: {string}")
print(f"upper string count: {d['upper']}")
print(f"lower string count: {d['lower']}")

org string: ORANGE or APPLE or BANNANA
upper string count: 18
lower string count: 4


In [0]:
'''
Unique numbers in list.
'''
mylist=[11,11,3,4,5,6,88,88,99,77,77]
uniqlist=list(set(mylist)) #[3, 4, 5, 6, 99, 11, 77, 88]


uniqlist=lambda mylist: list(set(mylist))
uniqlist(mylist)


[3, 4, 5, 6, 99, 11, 77, 88]

In [0]:
i=2
i*=3 
i+=3
i/=3
i**=2

i

9.0

[3, 6, 9]

In [0]:
'''
Multiply all numbers in a list
'''

a=[1,2,3]
listXed=[i*3 for i in a]

10

In [0]:
'''
Palindrome check
Write a Python function that checks 
whether a passed string is palindrome or not.

Note: A palindrome is word, phrase, or 
sequence that reads the same backward as forward,
 e.g., madam or nurses run.

string='madam'
string[:]==string[::-1]

string='nurses run'
string=string.replace(' ','')
string[:]==string[::-1]

'''
string='nurses run'

def Palindrome(string):
    string=string.replace(' ','')
    return string[:]==string[::-1]


print(f"{string} is a {Palindrome(string)}")

string[:]==string[::-1]

nurses run is a True


False

In [0]:
'''
Pangram check
A pangram is a sentence that contains all 
the letters of the English alphabet at least once. 
For example: The quick brown fox jumps over the lazy dog.
'''
aa="The quick brown fox jumps over the lazy dog"
a="apple"


AZlist=sorted(list(set(aa.replace(' ','').lower())))

lenAZlist=26

string="going to with a zebra is fun"
string="The quick brown fox jumps over the lazy dog"
def myfunc(string,AZlist):
    res=sorted(list(set(string.replace(' ','').lower())))
    return res==AZlist

myfunc(string,AZlist)



'abcdefghijklmnopqrstuvwxyz'

In [0]:
'''
Pangram check
A pangram is a sentence that contains all 
the letters of the English alphabet at least once. 
For example: The quick brown fox jumps over the lazy dog.

            import string
            alphabet=string.ascii_lowercase
            'abcdefghijklmnopqrstuvwxyz'

'''

import string

def ispangram(str1, alphabet=string.ascii_lowercase):  
    alphaset = set(alphabet)  
    return alphaset <= set(str1.lower())

# PRoject Assignment
Do this later!

Milestone Project 1: Full Walk-through Code Solution
Below is the filled in code that goes along with the complete walk-through video. Check out the corresponding lecture videos for more information on this code!

```
Address:

https://github.com/Pierian-Data/Complete-Python-3-Bootcamp/blob/master/04-Milestone%20Project%20-%201/03-Milestone%20Project%201%20-%20Complete%20Walkthrough%20Solution.ipynb

https://www.udemy.com/course/complete-python-bootcamp/learn/lecture/9442462#overview

```



#OOP
Object Oriented programming allows programmers to creat their own objects that have methods and attributes.

E.g. after defining a string,list, dict, or other objects, you were able to call the methods off of them with the method_name() syntax

OOP allows us to create code that is repeatable and organized


In [0]:
class Dog():
    # CLASS OBJECT ATTRIBUTE 
    # SAME FOR ANY INSTANCE OF CLASS
    # HENCE IT DOESN'T NEED attribute.self FORMAT
    species = 'mammal'

    # USER DEFINED ATTRIBUTES breed, name, spots
    def __init__(self,breed, name, spots):
        # ATTRIBUTES
        # TAKE IN ARGUMENTS
        # ASSIGN TO USER DEFINED ATTRIBUTES

        self.breed=breed #string
        self.name=name  #string
        self.spots=spots #boolen True/False


    # METHODS ARE FUNCTIONS ACTING ON AN OBJECT THROUGH THE USE OF SELF KEYWORD
    # METHODS ARE FUNCTIONS DEFINED INSIDE THE CLASS
    # THEY PERFORM OPPERATIONS USING ATTRIBUTES OF THE CREATED OBJECT

    def bark(self,number):
        print(f" woof! my name is {self.name} and my ID is {number}")

my_dog=Dog(breed='Lab', name= 'Doggie', spots=False)

type(my_dog)    #__main__.Dog
my_dog.species  #'mammal'
my_dog.bark(10)



 woof! my name is Doggie and my ID is 10


In [0]:
class Circle():
    # class object attribute 
    # is same on any instance of the class
    pi = 3.14
    def __init__(self,radius=1):
        self.radius=radius
    def get_area(self):
        return self.radius*self.pi*2
    
my_circle = Circle()
print('Default vals:',my_circle.pi,' and ',my_circle.radius)
print('area is: ', my_circle.get_area())

Default vals: 3.14  and  1
area is:  6.28


In [0]:
class Circle():
    # class object attribute 
    # is same on any instance of the class
    
    pi = 3.14
    
    def __init__(self,radius=1):
        self.radius=radius
        self.area=radius*self.pi*2

    def get_area(self):
        return self.radius*self.pi*2

my_circle=Circle()
my_circle.area

6.28

#Inheritence and Polymorphism

* Inheritence: is a way of forming new classes from existing already defined classes.(resduce complexity and reuse code

* newly formed classes can use the animal class in order to inherit some of its methods you may want to use again


```
Dog is inheriting or deriving from Animal class:

Dog(Animal)
```

Polymorphism: refers to the way in which different object classes can share the same method name, and those methods can be called from the same place eventhough a variety of different objects might be passed in.




#Inheritence

In [0]:
class Animal():
    def __init__(self):
        print("Animal created")
    def who_am_I(self):
        print("I am an animal")
    def eat(self):
        print("I am eating")


my_animal=Animal()
my_animal.who_am_I()
my_animal.eat()



Animal created
I am an animal
I am eating


In [0]:

class Dog(Animal):
    #creating an instance of the Dog class
    def __init__(self):
        #creating an instance of the Animal class
        Animal.__init__(self)
        print("Dog created")

    #overwrite inherited methods from base class if needed
    def who_am_I(self):
        print("I am overwritten version of the old dog from inhertence from Animal")
    
    #overwrite inherited methods from base class if needed
    def eat(self):
        print("I am eating a delicious food")
        
    #addon to methods from base class
    def bark(self):
        print("Woof!")

mydog = Dog()
mydog.eat()
mydog.who_am_I()
mydog.bark()



Animal created
Dog created
I am eating a delicious food
I am overwritten version of the old dog from inhertence from Animal
Woof!


# Polymorphism

In [0]:
class Dog():
    def __init__(self,name):
        self.name=name
    def speak(self):
        return self.name+" says woof!"

class Cat():
    def __init__(self,name):
        self.name=name
    def speak(self):
        return self.name+" says meow!"


niko = Dog("Niko")
felix = Cat("Felix")

print(niko.speak())
print(felix.speak())

Niko says woof!
Felix says meow!


<class '__main__.Dog'>
Niko says woof!
<class '__main__.Cat'>
Felix says meow!


In [0]:
################ polymorphism using function ########################
# pet_speak function is calling .speak() method
# it doesn't know which of dog or cat class
# is going to be passed on by the user.

# we get object specific results by 
# calling same function (e.g pet_speak)
def pet_speak(pet):
    print(pet.speak())

pet_speak(niko)

################# polymorphism using for loop #######################
for pet in [niko, felix]:
    print(type(pet))
    print(pet.speak())


Niko says woof!
<class '__main__.Dog'>
Niko says woof!
<class '__main__.Cat'>
Felix says meow!


# Abstract class

* when we have a class that never expects to be instantiated
instead it is only meant to be used * as a base class


In [0]:
class Animal():
    def __init__(self,name):
        self.name=name
    def speak(self):
        raise NotImplementedError("subclass must be implement this abstract method")

class Dog(Animal):
    def speak(self):
        return self.name+" says woof!" 

class Cat(Animal):
    def speak(self):
        return self.name+" says meaw!" 

fido = Dog("Fido")
sphinx = Cat("Sphinx") 

print(fido.speak())
print(sphinx.speak())

Fido says woof!
Sphinx says meaw!


# Magic methods
Use builtin python functions like (len and print) with my own user defined objects.


In [0]:
class Book():
    def __init__(self, title, author, pages):
        self.title = title
        self.author = author
        self.pages=pages

    # MOST POPULAR MAGIC/DUNDER METHODS
    def __str__(self): #print function extracts string from here for printing
        return f"{self.title} by {self.author}"

    def __len__(self): #show len when we call len
        return self.pages
    
    def __del__(self): # show the text bellow when u dlete the instance
        print("Book has been deleted")


b = Book('Python rocks','Maryam',200)
print(b)
len(b)
del(b) # can type: del b


Python rocks by Maryam
Book has been deleted


In [0]:
# Fill in the Line class methods to accept coordinates as a pair 
# of tuples and return the slope and distance of the line.
class Line(object):
    
    def __init__(self,coor1,coor2):
        self.coor1 = coor1
        self.coor2 = coor2
    
    def distance(self):
        x1,y1 = self.coor1 # x,y = tuple unpacking
        x2,y2 = self.coor2
        #2 decimal points
        return f"{((x2-x1)**2 + (y2-y1)**2)**0.5:0.2f}"
    
    def slope(self):
        x1,y1 = self.coor1
        x2,y2 = self.coor2
        return (y2-y1)/(x2-x1)



coordinate1 = (3,2)
coordinate2 = (8,10)

li = Line(coordinate1,coordinate2)

li.distance()

#li.slope()


'9.43'

In [0]:
class Cylinder:
    
    def __init__(self,height=1,radius=1):
        self.height = height
        self.radius = radius
        
    def volume(self):
        return self.height*3.14*(self.radius)**2
    
    def surface_area(self):
        top = 3.14 * (self.radius)**2
        return (2*top) + (2*3.14*self.radius*self.height)

c = Cylinder(2,3)

c.volume() #56.52

c.surface_area()# 94.2

94.2

In [0]:
class Account:
    def __init__(self,owner,balance=0):
        self.owner = owner
        self.balance = balance
        
    def __str__(self):
        return f'Account owner:   {self.owner}\nAccount balance: ${self.balance}'
        
    def deposit(self,dep_amt):
        self.balance += dep_amt
        print('Deposit Accepted')
        
    def withdraw(self,wd_amt):
        if self.balance >= wd_amt:
            self.balance -= wd_amt
            print('Withdrawal Accepted')
        else:
            print('Funds Unavailable!')

# 1. Instantiate the class
acct1 = Account('Jose',100)

# 2. Print the object
print(acct1)
Account owner:   Jose #Account balance: $100

# 3. Show the account owner attribute
acct1.owner #'Jose'

# 4. Show the account balance attribute
acct1.balance #100

# 5. Make a series of deposits and withdrawals
acct1.deposit(50)

acct1.withdraw(75) #Withdrawal Accepted

# 6. Make a withdrawal that exceeds the available balance
acct1.withdraw(500)

# Create your own modules and packages and subpackages

* Modules are just .py scripts that you call in another .py script
* packages are collection of modules.

* if you have a large script you are not going to put everything in a giant .py file. Instead you split it up among different .py files which means you split it among different modules. 
    

```
#mymodule.py and mymodule2 are both under same directory and contain my_func and my_func2 functions correspondingly

from mymodule import my_func
from mymodule2 import my_func2 as mf

```



* At some point having too many modules is not easy to handle and you can use a bunch of modules to create different packages.
    * create a 'directory' (e.g "MyMainPackageDir") and inside that you can have 'subdirectories' e.g "SubPackageDir").
    * create a blank files with '__init__.py' name in all 'directory' and 'subdirectories' so when python goes searching inside these packages it understand that it is not a normal directory and it is an actual package

```

from MyMainPackageDir import some_main_script

from MyMainPackageDir.SubPackageDir import mysubscript

some_main_script.my_func_definition()

mysubscript.my_func_definition()

```







In [0]:
!pwd

/content


In [0]:

with open('mymodule.py','w') as myfile:
    myfile.write("def my_func():\n\tprint('Hey I am in mymodule')\n")

with open('mymodule2.py','w') as myfile:
    myfile.write("""
def my_func2():
    print('Hey I am in mymodule 2')""")


with open('myprogram.py','w') as myfile:
    myfile.write("""
from mymodule import my_func
from mymodule2 import my_func2 as mf
my_func()
mf()
""")




In [0]:
!cat myprogram.py
!cat myprogram2.py
!cat mymodule.py

!python myprogram.py



from mymodule import my_func
from mymodule2 import my_func2 as mf
my_func()
mf()

def my_func():
    print('Hey I am in mymodule')def my_func():
	print('Hey I am in mymodule')
Hey I am in mymodule
Hey I am in mymodule 2


In [0]:
print("""
hi
*    holla
heyyy
*            hollaaaa
""")


hi
*    holla
heyyy
*            hollaaaa



# \_\_name\_\_ and \_\_main\_\_

* Sometimes you see at the bottom of the code 

    * if \_\_name\_\_ == \_\_main\_\_

* sometime when you are importing from a module, you want to know whether a module function is being used as an import, or if you are using the original.py file of that module
    * on desktop create one.py and two.py
    * when you call a ".py" file all the code will gets run
    * in python there is a builtin variabe called "\_\_name\_\"
    * This variable "\_\_name\_\" gets assigned different names
    * when you run a file.py directly it assignes 
"\_\_name\_\"="\_\_main\_\_"


```
# This is formatted as code

if __name__=__main__

```




Sometimes when you are importing from a module, you would like to know whether
a modules function is being used as an import, or if you are using the original
.py file of that module. In this case we can use the:

      if __name__ == "__main__":

line to determine this. For example:

When your script is run by passing it as a command to the Python interpreter:

    python myscript.py

all of the code that is at indentation level 0 gets executed. Functions and
classes that are defined are, well, defined, but none of their code gets ran.
Unlike other languages, there's no main() function that gets run automatically
- the main() function is implicitly all the code at the top level.

In this case, the top-level code is an if block.  __name__ is a built-in variable
 which evaluate to the name of the current module. However, if a module is being
 run directly (as in myscript.py above), then __name__ instead is set to the
 string "__main__". Thus, you can test whether your script is being run directly
  or being imported by something else by testing

    if __name__ == "__main__":
        ...

If that code is being imported into another module, the various function and
class definitions will be imported, but the main() code won't get run. As a
basic example, consider the following two scripts:

    # file one.py
    def func():
        print("func() in one.py")

    print("top-level in one.py")

    if __name__ == "__main__":
        print("one.py is being run directly")
    else:
        print("one.py is being imported into another module")

and then:

    # file two.py
    import one

    print("top-level in two.py")
    one.func()

    if __name__ == "__main__":
        print("two.py is being run directly")
    else:
        print("two.py is being imported into another module")

Now, if you invoke the interpreter as

    python one.py

The output will be

    top-level in one.py

one.py is being run directly
If you run two.py instead:

    python two.py

You get

  top-level in one.py
  one.py is being imported into another module
  top-level in two.py
  func() in one.py
  two.py is being run directly
  
Thus, when module one gets loaded, its __name__ equals "one" instead of __main__.**bold text**

#Creating a Decorator

When you want to add more new capabilities to a function, adding extra code to old code or creating new codes are NOT good ways to proceed. Because you don't either have capability of calling the old function or you had to create the old function over again and add more new code and functionality.

Thats when decorateors come to the picture: Python has decorators that allow you to attach new extra functionality to an already existing function. They use "@" operator and are then placed on top of the original function.


* easily add on extra functionality with a decoder
```
    @some_decoder

    def simple_func():
    # Do simple stuff
        return something

```

#RETURN A FUNC WITHIN a FUNC

In [0]:
def hello(name='Jose'):
    print("the hello function has been executed")
    def greet():
        return'\t this is greet() func inside hello!'

    def welcome():
        return'\t this is welcome() func inside hello!'

    print('I am going to return a function')
    if name =='Jose':
        return greet #  ATTENTION NO PARANTHESIS greet()
    else:
        return welcome #  ATTENTION NO PARANTHESIS

# DONT DO THE JOB - YOU NEED IT CALLED LIKE A FUNCTION
print(hello('Jose'))

# the hello function has been executed
# I am going to return a function
# <function hello.<locals>.greet at 0x7f61abbfdf28>

# DOES THE JOB
newFunc=hello('Jose')
print(newFunc())
# the hello function has been executed
# I am going to return a function
# this is greet() func inside hello!

the hello function has been executed
I am going to return a function
<function hello.<locals>.greet at 0x7f61abbfdf28>
the hello function has been executed
I am going to return a function
	 this is greet() func inside hello!


In [0]:
def cool():
    def super_cool():
        return 'I am very cool '
    return super_cool

some_func = cool()
some_func()

'I am very cool '

In [0]:
def new_decorator(original_func):
        def wrap_func():
            print('some extra code added before original func')
            original_func()
            print('some extra code added after original func')
        return wrap_func


In [0]:
def func_needs_decorator():
    print("I want to be decorated!")

decorated_func = new_decorator(func_needs_decorator)

decorated_func()

# some extra code added before original func
# I want to be decorated!
# some extra code added after original func

some extra code added before original func
I want to be decorated!
some extra code added after original func


The above was complicated...Now lets use decorator:


In [0]:
def new_decoratorr(original_func):
        def wrap_func():
            print('*some extra code added before original func')
            original_func()
            print('*some extra code added after original func')
        return wrap_func


@new_decoratorr
def func_needs_decorator():
    print("*I want to be decorated!")

func_needs_decorator()

*some extra code added before original func
*I want to be decorated!
*some extra code added after original func



In the previous example we actually manually created a Decorator. Here we will modify it to make its use case clear:

```
def new_decorator(func):

    def wrap_func():
        print("Code would be here, before executing the func")

        func()

        print("Code here will execute after the func()")

    return wrap_func

def func_needs_decorator():
    print("This function is in need of a Decorator")

    ```

    So what just happened here? A decorator simply wrapped the function and modified its behavior. Now let's understand how we can rewrite this code using the @ symbol, which is what Python uses for Decorators:



```
@new_decorator
def func_needs_decorator():
    print("This function is in need of a Decorator")

func_needs_decorator()

```

Great! You've now built a Decorator manually and then saw how we can use the @ symbol in Python to automate this and clean our code. You'll run into Decorators a lot if you begin using Python for Web Development, such as Flask or Django!



# Errors and Exception Handling

[link text](https://docs.python.org/3/library/exceptions.html)


If you execute this you get an error and you can't run anything after
```
In [1]:
print('Hello)
  File "<ipython-input-1-db8c9988558c>", line 1
    print('Hello)




SyntaxError: EOL while scanning string literal
```

Note how we get a SyntaxError, with the further description that it was an EOL (End of Line Error) while scanning the string literal. This is specific enough for us to see that we forgot a single quote at the end of the line. Understanding these various error types will help you debug your code much faster.

This type of error and description is known as an Exception. Even if a statement or expression is syntactically correct, it may cause an error when an attempt is made to execute it. Errors detected during execution are called exceptions and are not unconditionally fatal.

You can check out the full list of built-in exceptions here. Now let's learn how to handle errors and exceptions in our own code.

# try (1) , except (2a), else (2a) , finally (3)
The basic terminology and syntax used to handle errors in Python are the try and except statements. The code which can cause an exception to occur is put in the try block and the handling of the exception is then implemented in the except block of code. The syntax follows:

```
try:
   You do your operations here...
   ...
except ExceptionI:
   If there is ExceptionI, then execute this block.
except ExceptionII:
   If there is ExceptionII, then execute this block.
   ...
else:
   If there is no exception then execute this block. 
   ```

We can also just check for any exception with just using except: To get a better understanding of all this let's check out an example: We will look at some code that opens and writes a file:

```
try:
    f = open('testfile','w')
    f.write('Test write this')
except IOError:
    # This will only check for an IOError exception and then execute this print statement
    print("Error: Could not find file or read data")
else:
    print("Content written successfully")
    f.close()

    ```

    ```
Content written successfully
Now let's see what would happen if we did not have write permission (opening only with 'r'):
```



```
try:
    f = open('testfile','r')
    f.write('Test write this')
except IOError:
    # This will only check for an IOError exception and then execute this print statement
    print("Error: Could not find file or read data")
else:
    print("Content written successfully")
    f.close()

    ```

Error: Could not find file or read data
Great! Notice how we only printed a statement! The code still ran and we were able to continue doing actions and running code blocks. This is extremely useful when you have to account for possible input errors in your code. You can be prepared for the error and keep running code, instead of your code just breaking as we saw above.

We could have also just said except: if we weren't sure what exception would occur. For example:
```
try:
    f = open('testfile','r')
    f.write('Test write this')
except:
    # This will check for any exception and then execute this print statement
    print("Error: Could not find file or read data")
else:
    print("Content written successfully")
    f.close()
    ```
    ```
Error: Could not find file or read data
Great! Now we don't actually need to memorize that list of exception types! Now what if we kept wanting to run code after the exception occurred? This is where finally comes in.

finally
The finally: block of code will always be run regardless if there was an exception in the try code block. The syntax is:
```

```
try:
   Code block here
   ...
   Due to any exception, this code may be skipped!
finally:
   This code block would always be executed.
```
For example:

```
try:
    f = open("testfile", "w")
    f.write("Test write statement")
    f.close()
finally:
    print("Always execute finally code blocks")
Always execute finally code blocks
We can use this in conjunction with except. Let's see a new example that will take into account a user providing the wrong input:
```
```
def askint():
    try:
        val = int(input("Please enter an integer: "))
    except:
        print("Looks like you did not enter an integer!")

    finally:
        print("Finally, I executed!")
    print(val)
```
askint()

Please enter an integer: 5
Finally, I executed!
5

askint()
Please enter an integer: five
Looks like you did not enter an integer!
Finally, I executed!
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-8-cc291aa76c10> in <module>()
----> 1 askint()

<ipython-input-6-c97dd1c75d24> in askint()
      7     finally:
      8         print("Finally, I executed!")
----> 9     print(val)

UnboundLocalError: local variable 'val' referenced before assignment
Notice how we got an error when trying to print val (because it was never properly assigned). Let's remedy this by asking the user and checking to make sure the input type is an integer:

```
def askint():
    try:
        val = int(input("Please enter an integer: "))
    except:
        print("Looks like you did not enter an integer!")
        val = int(input("Try again-Please enter an integer: "))
    finally:
        print("Finally, I executed!")
    print(val)
```

askint()
Please enter an integer: five
Looks like you did not enter an integer!
Try again-Please enter an integer: four
Finally, I executed!
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-92b5f751eb01> in askint()
      2     try:
----> 3         val = int(input("Please enter an integer: "))
      4     except:

ValueError: invalid literal for int() with base 10: 'five'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-10-cc291aa76c10> in <module>()
----> 1 askint()

<ipython-input-9-92b5f751eb01> in askint()
      4     except:
      5         print("Looks like you did not enter an integer!")
----> 6         val = int(input("Try again-Please enter an integer: "))
      7     finally:
      8         print("Finally, I executed!")

ValueError: invalid literal for int() with base 10: 'four'
Hmmm...that only did one check. How can we continually keep checking? We can use a while loop!

```
def askint():
    while True:
        try:
            val = int(input("Please enter an integer: "))
        except:
            print("Looks like you did not enter an integer!")
            continue
        else:
            print("Yep that's an integer!")
            break
        finally:
            print("Finally, I executed!")
        print(val)
```
askint()
Please enter an integer: five
Looks like you did not enter an integer!
Finally, I executed!
Please enter an integer: four
Looks like you did not enter an integer!
Finally, I executed!
Please enter an integer: 3
Yep that's an integer!
Finally, I executed!
So why did our function print "Finally, I executed!" after each trial, yet it never printed val itself? This is because with a try/except/finally clause, any continue or break statements are reserved until after the try clause is completed. This means that even though a successful input of 3 brought us to the else: block, and a break statement was thrown, the try clause continued through to finally: before breaking out of the while loop. And since print(val) was outside the try clause, the break statement prevented it from running.

Let's make one final adjustment:

```
def askint():
    while True:
        try:
            val = int(input("Please enter an integer: "))
        except:
            print("Looks like you did not enter an integer!")
            continue
        else:
            print("Yep that's an integer!")
            print(val)
            break
        finally:
            print("Finally, I executed!")
            ```

askint()
Please enter an integer: six
Looks like you did not enter an integer!
Finally, I executed!
Please enter an integer: 6
Yep that's an integer!
6
Finally, I executed!
Great! Now you know how to handle errors and exceptions in Python with the try, except, else, and finally notation!

In [0]:
try:
    for i in ['a','b','c']:
        print(i**2)
except TypeError:
    print("Type error")

Type error


In [0]:
try:
    z=5/0
except TypeError:
    print("Type error")
except:
    print("ZeroDivisionError")

ZeroDivisionError


In [0]:
# try except else
while True:
    try:
        a=int(input("Please input int: "))
        
    except:
        print("Type error! please type int!")
        continue #optional
    else:
        print(f"thank you")
        break
print(f"here is sqr: {a**2}, thank you")


Please input int: g
Type error! please type int!
Please input int: 12
thank you
here is sqr: 144, thank you


In [0]:
def ask():
    waiting=True
    while waiting:
        try:
            a=int(input("Please input int: "))
            
        except:
            print("Type error! please type int!")
            continue #optional
        else:
            print(f"thank you")
            waiting=False
    print(f"here is sqr: {a**2}, thank you")

ask()

Type error! please type int!
Please input int: 1
thank you
here is sqr: 1, thank you


##Unit Testing
Equally important as writing good code is writing good tests. Better to find bugs yourself than have them reported to you by end users!

For this section we'll be working with files outside the notebook. We'll save our code to a .py file, and then save our test script to another .py file. Normally we would code these files using a text editor like Brackets or Atom, or inside an IDE like Spyder or Pycharm. But, since we're here, let's use Jupyter!

Recall that with some IPython magic we can write the contents of a cell to a file using %%writefile.
Something we haven't seen yet; you can run terminal commands from a jupyter cell using !


##Testing tools
There are dozens of good testing libraries out there. Most are third-party packages that require an install, such as:

* pylint: cheks your code for possible error and styling 
* pyflakes
* pep8 : set of style convention rules in python

These are simple tools that merely look at your code, and they'll tell you if there are style issues or simple problems like variable names being called before assignment.

A far better way to test your code is to write tests that send sample data to your program, and compare what's returned to a desired outcome.
Two such tools are available from the standard library:

* unittest
* doctest

Let's look at pylint first, then we'll do some heavier lifting with unittest.

##pylint
pylint tests for style as well as some very basic program logic.

First, if you don't have it already (and you probably do, as it's part of the Anaconda distribution), you should install pylint.
Once that's done feel free to comment out the cell, you won't need it anymore.


```

! pip install pylint

```
Let's save a very simple script:

# This is formatted as code
```
%%writefile simple1.py
a = 1
b = 2
print(a)
print(B)
# This is formatted as code
```

Overwriting simple1.py
Now let's check it using pylint
```
! pylint simple1.py
```

In [0]:
! pip install pylint

     |████████████████████████████████| 307kB 2.8MB/s 
     |████████████████████████████████| 51kB 7.1MB/s 
     |████████████████████████████████| 215kB 38.2MB/s 
     |████████████████████████████████| 747kB 39.0MB/s 
     |████████████████████████████████| 61kB 8.4MB/s 


In [0]:

%%writefile simple1.py
a = 1
b = 2
print(a)
print(B)

Writing simple1.py


In [0]:
! pylint simple1.py

************* Module simple1
simple1.py:4:0: C0304: Final newline missing (missing-final-newline)
simple1.py:1:0: C0114: Missing module docstring (missing-module-docstring)
simple1.py:1:0: C0103: Constant name "a" doesn't conform to UPPER_CASE naming style (invalid-name)
simple1.py:2:0: C0103: Constant name "b" doesn't conform to UPPER_CASE naming style (invalid-name)
simple1.py:4:6: E0602: Undefined variable 'B' (undefined-variable)

-------------------------------------
Your code has been rated at -12.50/10



Pylint first lists some styling issues - it would like to see an extra newline at the end, modules and function definitions should have descriptive docstrings, and single characters are a poor choice for variable names.

More importantly, however, pylint identified an error in the program - a variable called before assignment. This needs fixing.

Note that pylint scored our program a negative 12.5 out of 10. Let's try to improve that!

In [0]:
%%writefile simple2.py
"""
A very simple script.
"""

def myfunc():
    """
    An extremely simple function.
    """
    first = 1
    second = 2
    print(first)
    print(second)

myfunc()

In [0]:
! pylint simple2.py

************* Module simple2.py
simple2.py:1:0: F0001: No module named simple2.py (fatal)


Much better! Our score climbed to 8.33 out of 10. Unfortunately, the final newline has to do with how jupyter writes to a file, and there's not much we can do about that here. Still, pylint helped us troubleshoot some of our problems. But what if the problem was more complex?

In [0]:
%%writefile simple2.py
"""
A very simple script.
"""

def myfunc():
    """
    An extremely simple function.
    """
    first = 1
    second = 2
    print(first)
    print('second')

myfunc()

pylint tells us there's an unused variable in line 10, but it doesn't know that we might get an unexpected output from line 12! For this we need a more robust set of tools. That's where unittest comes in.


## unittest
unittest lets you write your own test programs. The goal is to send a specific set of data to your program, and analyze the returned results against an expected result.

Let's generate a simple script that capitalizes words in a given string. We'll call it cap.py.


In [0]:

%%writefile cap.py
def cap_text(text):
    return text.capitalize()

Writing cap.py


Overwriting cap.py
Now we'll write a test script. We can call it whatever we want, but test_cap.py seems an obvious choice.

When writing test functions, it's best to go from simple to complex, as each function will be run in order. Here we'll test simple, one-word strings, followed by a test of multiple word strings.

In [0]:
%%writefile test_cap.py # tests cap.py
import unittest
import cap #calls cap.py

# we inherit from unittest.TestCase
class TestCap(unittest.TestCase):
    
    def test_one_word(self):
        text = 'python'
        result = cap.cap_text(text)
        self.assertEqual(result, 'Python')
        
    def test_multiple_words(self):
        text = 'monty python'
        result = cap.cap_text(text)
        self.assertEqual(result, 'Monty Python')
        
if __name__ == '__main__':
    unittest.main()



UsageError: unrecognized arguments: # tests cap.py


What happened? It turns out that the .capitalize() method only capitalizes the first letter of the first word in a string. Doing a little research on string methods, we find that .title() might give us what we want.

# Iterators and Generators
**Generators are best for calculating large sets of results (particularly in calculations that involve loops themselves) in cases where we don’t want to allocate the memory for all of the results at the same time.**

There are differences between iteration and generation in Python and how to construct our own Generators with the yield statement. Generators allow us to generate as we go along, instead of holding everything in memory.

* built-in Python functions like:
 * range(), map() and filter().


We created functions with def and the return statement. Generator functions allow us to write a function that can send back a value and then later resume to pick up where it left off. This type of function is a generator in Python, allowing us to generate a sequence of values over time. The main difference in syntax will be the use of a yield statement.

In most aspects, a generator function will appear very similar to a normal function. The main difference is when a generator function is compiled they become an object that supports an iteration protocol. That means when they are called in your code they don't actually return a value and then exit. Instead, generator functions will automatically suspend and resume their execution and state around the last point of value generation. The main advantage here is that instead of having to compute an entire series of values up front, the generator computes one value and then suspends its activity awaiting the next instruction. This feature is known as state suspension.




In [0]:
# NOT Memory Efficient!
# because we are keepin an entire result[] list in  memory
def creat_cubes(n):
    result=[]
    for x in range(n):
        result.append(x**3)
    return result



for x in creat_cubes(10):
    print(x)

0
1
8
27
64
125
216
343
512
729


In [0]:
#USE YIELD TO BE MIORE MEMORY EFFICIENT
def creat_cubes(n):

    for x in range(n):
        yield x**3

for x in creat_cubes(10):
    print(x)

# if you needed a list you can still make it
list(creat_cubes(10))

0
1
8
27
64
125
216
343
512
729


LIST and/or FOR LOOP helps seeing the output of the GENERATOR without having to type NEXT() statement to itterate through it.

In [0]:
def sqgenerator(N):
    for i in range(N):
        yield i**2

for x in sqgenerator(10):
    print(x)


list(sqgenerator(10))

0
1
4
9
16
25
36
49
64
81


[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

In [0]:
# make generator for random numbers
import random
random.randint(1,10) #randint(a,b) range[a,b]


'''
Create a generator that yields "N" random numbers between a,b
'''
def rand_num(a,b,n):
    for i in range(n):
        yield random.randint(a,b)
for num in rand_num(1,10,12):
    print(f"{num} \n")


8 

5 

5 

7 

10 

9 

5 

9 

4 

2 

3 

8 



In [0]:
#convert string from itterable to an itterator
s="hello"
s=iter(s)
print(next(s))
print(next(s))
print(next(s))
print(next(s))
print(next(s))

h
e
l
l
o


# GenComp: Generator compression

It is similar to list comprehension but instead of creating a list in memory it is going to generate it and all you need to do is to  itterate through your generator and not save all of this into the memory!




In [0]:
my_list=[1,2,3,4,5]

gencomp = (item for item in my_list if item >3 )

for item in gencomp:
    print(item)

4
5


# Advanced Python Modules

















### Collections Module

The collections module is a built-in module that implements specialized container data types providing alternatives to Python’s general purpose built-in containers. We've already gone over the basics: dict, list, set, and tuple.

Now we'll learn about the alternatives that the collections module provides.

In [0]:
#Counter
from collections import Counter
# Bellow it counts number of repeated in a list and acts as a dictionary
L=[1,1,1,2,2,2,333,3,3,3,]
Counter(L)
#output: Counter({1: 3, 2: 3, 3: 3, 333: 1})

# Bellow it counts number of repeated in a string and acts as a dictionary

s='asshehhrjekksbbsbbbebe'
Counter(s)
#output: Counter({'a': 1, 'b': 6, 'e': 4, 'h': 3, 'j': 1, 'k': 2, 'r': 1, 's': 4})


sentence="How many times words repeat in this sentece? word word show!"

words=sentence.split()
Counter(words)
# Counter({'How': 1,
        #  'in': 1,
        #  'many': 1,
        #  'repeat': 1,
        #  'sentece?': 1,
        #  'show!': 1,
        #  'this': 1,
        #  'times': 1,
        #  'word': 2,
        #  'words': 1})


Counter({'How': 1,
         'in': 1,
         'many': 1,
         'repeat': 1,
         'sentece?': 1,
         'show!': 1,
         'this': 1,
         'times': 1,
         'word': 2,
         'words': 1})

In [0]:
sentence="How many times words repeat in this sentece? word word show!"

words=sentence.split()
c = Counter(words)
c.most_common(3) #top three most common
[('word', 2), ('How', 1), ('many', 1)]
sum(c.values)

[('word', 2), ('How', 1), ('many', 1)]




**Common patterns when using the Counter() object**

Counter is a dictionary so you can use all these that can be used in dicts


```
sum(c.values())                 # total of all counts
c.clear()                       # 

1.   List item
2.   List item

reset all counts
list(c)                         # list unique elements
set(c)                          # convert to a set
dict(c)                         # convert to a regular dictionary
c.items()                       # convert to a list of (elem, cnt) pairs
Counter(dict(list_of_pairs))    # convert from a list of (elem, cnt) pairs
c.most_common()[:-n-1:-1]       # n least common elements
c += Counter()                  # remove zero and negative counts


```

#defaultdict
defaultdict is a dictionary-like object which provides all methods provided by a dictionary but takes a first argument (default_factory) as a default data type for the dictionary. Using defaultdict is faster than doing the same using dict.set_default method.

Unline norml dictionaries, the default dict doesn't give you error whenyou request for a key not present in your dictionary

In [0]:
from collections import defaultdict

## Normal dict
# d={'k1':1}
# d['k2'] : Error!: d['k2']=

d = defaultdict(object)
d["one"]
d["two"]=3
d
                                                                                           

defaultdict(object, {'one': <object at 0x7f61bb428e30>, 'two': 3})

#Ordered dict
A normal Dict is only a mapping and doesn't retain the order of the keys

In [0]:
from collections import OrderedDict
d={}
d['a'],d['b'],d['c'], d['d'],d['e'] = 1,2,3,4,5


a 1
b 2
c 3
d 4
e 5


True

In [0]:
from collections import OrderedDict

d = OrderedDict()
d['a'],d['b'],d['c'], d['d'],d['e'] = 1,2,3,4,5

for k,v in d.items(): #the order is retained
    print(k,v)



a 1
b 2
c 3
d 4
e 5


In [0]:
############ Order doesn't matter - Default dict####################
from collections import defaultdict

d1,d2= defaultdict(),defaultdict()

d1['a']=2
d1['b']=1

d2['b']=1
d2['a']=2

d1==d2 #True

############ Order doesn't matter - Normal dict ####################
#{'a': 1, 'b': 2, 'c': 3}
for k,v in d.items():
    print(k,v)

d1,d2={},{}
d1['a']=1
d1['b']=2

d2['b']=2
d2['a']=1

d1==d2 #True

############ Order of filling dictionary MATTERS! OrderedDict ####################

from collections import OrderedDict

d1,d2= OrderedDict(),OrderedDict()

d1['a']=2
d1['b']=1

d2['b']=1
d2['a']=2

d1==d2 #FALSE!

a 1
b 2
c 3
d 4
e 5


False

#named tuple

it assigns names as well as numerical index to eaach member of the tuple


In [0]:
# norml tuple
t=(1,2,3)
t[0]

1

In [0]:
from collections import namedtuple

Dog = namedtuple('Dog','age breed name')
Cat = namedtuple('Cat','fur claws name')

sam =Dog(age=2, breed='Lab', name='Sammy')
sam.age # 2
sam.breed #Lab
sam.name  #sammy
sam.index('Sammy') #2
sam.index('Lab') #1
sam.index(2) #0
sam.count('2') #0
sam.count(2) #1
sam.count('Lab') #1


c=Cat(fur= 'Fuzzy', claws=False, name='Kitty')
c.name # 'Kitty'
c[2] #'Kitty'







'Kitty'

#datetime module


In [0]:
import datetime

t = datetime.time(4, 20, 1)

# Let's show the different components
print(t)
print('hour  :', t.hour)
print('minute:', t.minute)
print('second:', t.second)
print('microsecond:', t.microsecond)
print('tzinfo:', t.tzinfo)

print('Earliest  :', datetime.time.min)
print('Latest    :', datetime.time.max)
print('Resolution:', datetime.time.resolution)

05:25:02
00:00:00
23:59:59.999999
0:00:00.000001
2020-02-10
2020
0001-01-01
9999-12-31
1 day, 0:00:00


In [0]:
today = datetime.date.today()
print(today)
print('ctime:', today.ctime())
print('tuple:', today.timetuple())
print('ordinal:', today.toordinal())
print('Year :', today.year)
print('Month:', today.month)
print('Day  :', today.day)


print('Earliest  :', datetime.date.min)
print('Latest    :', datetime.date.max)
print('Resolution:', datetime.date.resolution)


In [0]:
d1 = datetime.date(2015, 3, 11)
print('d1:', d1)

d2 = d1.replace(year=1990)
print('d2:', d2)


print('d1-d2', d1-d2)

#list min max sum

In [0]:
l=[11,3,2,30]
min(l)
sum(l)
max(l)

30

# Python Debugger (PDB)

You've probably used a variety of print statements to try to find errors in your code. A better way of doing this is by using Python's built-in debugger module (pdb). The pdb module implements an interactive debugging environment for Python programs. It includes features to let you pause your program, look at the values of variables, and watch program execution step-by-step, so you can understand what your program actually does and find bugs in the logic.

This is a bit difficult to show since it requires creating an error on purpose, but hopefully this simple example illustrates the power of the pdb module.
Note: Keep in mind it would be pretty unusual to use pdb in an iPython Notebook setting.






In [0]:
import pdb

x = [1,3,4]
y = 2
z = 3

result = y + z
print(result)

pdb.set_trace() # put it before error, press "q" to quit 

#Error result2 = y+x
#print(result2)



5
--Return--
> <ipython-input-137-e9cff1a5e62b>(10)<module>()->None
-> pdb.set_trace() # put it before error, press "q" to quit
(Pdb) q


BdbQuit: ignored

#timing your code - timeit


In [0]:
'''
creat this string and time it

0-1-2-3-...,99

'''

# For loop
timeit.timeit('"-".join(str(n) for n in range(100))', number=10000)
#0.21865416520477374
In [3]:
# List comprehension
timeit.timeit('"-".join([str(n) for n in range(100)])', number=10000)
# 0.19484614421698643
In [4]:
# Map()
timeit.timeit('"-".join(map(str, range(100)))', number=10000)
Out[4]:
0.15291817337139246

2.5783022020004864

Let's repeat the above examinations using iPython magic!


In [0]:
# This is formatted as code

%timeit "-".join(str(n) for n in range(100))
#20.4 µs ± 269 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%timeit "-".join([str(n) for n in range(100)])
#18.1 µs ± 56.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit "-".join(map(str, range(100)))
#14.4 µs ± 64.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


10000 loops, best of 3: 22.9 µs per loop
10000 loops, best of 3: 20.6 µs per loop
100000 loops, best of 3: 14.4 µs per loop


## Regular Expressions regexp regex

**************************

[link for more](https://www.tutorialspoint.com/python/python_reg_expressions.htm)

[full doc](https://docs.python.org/3/library/re.html#regular-expression-syntax)

### import re
* re.findall(A, B) | Matches all instances of an expression A in a string B and returns them in a list.

* re.search(A, B) | Matches the first instance of an expression A in a string B, and returns it as a re match object.

* re.split(A, B) | Split a string B into a list using the delimiter A.

* re.sub(A, B, C) | Replace A with B in the string C.

****************************



```
.	Any character except newline
a	The character a
ab	The string ab
a|b	a or b
a*	0 or more a's
\	Escapes a special character


*	0 or more
+	1 or more
?	0 or 1
{2}	Exactly 2
{2, 5}	Between 2 and 5
{2,}	2 or more
(,5}	Up to 5


ab-d]	One character of: a, b, c, d
[^ab-d]	One character except: a, b, c, d
[\b]	Backspace character
\d	One digit
\D	One non-digit
\s	One whitespace
\S	One non-whitespace
\w	One word character
\W	One non-word character


^	Start of string
\A	Start of string, ignores m flag
$	End of string
\Z	End of string, ignores m flag
\b	Word boundary
\B	Non-word boundary
(?=...)	Positive lookahead
(?!...)	Negative lookahead
(?<=...)	Positive lookbehind
(?<!...)	Negative lookbehind
(?()|)	Conditional


i	Ignore case
m	^ and $ match start and end of line
s	. matches newline as well
x	Allow spaces and comments


\n	Newline
\r	Carriage return
\t	Tab

```
*****************************

```
examples
'abc'                 # Matches the literal 'abc'
r'\w+'                # Matches a word
r'\d+\s\w+'           # Matches '1555 dogs'
r'\d+-\d+-\d+\s.+'    # Matches '2005-12-05 Jons birthday'
r'\s+'                # Matches any number of whitespace

[a-z0-9] | Matches characters from a to z and also from 0 to 9.
[a-z] | Matches any alphabet from a to z.
[amk] | Matches either a, m, or k. It does not match amk.
```
Regular expressions are text-matching 
patterns described with a formal syntax. You'll often hear regular expressions referred to as 'regex' or 'regexp' in conversation. Regular expressions can include a variety of rules, from finding repetition, to text-matching, and much more. As you advance in Python you'll see that a lot of your parsing problems can be solved with regular expressions (they're also a common interview question!).

If you're familiar with Perl, you'll notice that the syntax for regular expressions are very similar in Python. We will be using the re module with Python for this lecture.

###Searching for Patterns in Text
One of the most common uses for the re module is for finding patterns in text. Let's do a quick example of using the search method in the re module to find some text:

In [0]:
import re

match=re.search('yam',"I am Maryam, Je sue Maryam ")
print(match)

# Show start of match
match.start() #8

# Show end
match.end() #11

<_sre.SRE_Match object; span=(8, 11), match='yam'>


11

In [0]:
import re


re.sub()
# List of patterns to search for
patterns = ['term1', 'term2']

# Text to parse
text = 'This is a string with term1, but it does not have the other term.'

for pattern in patterns:
    print('Searching for "%s" in:\n "%s"\n' %(pattern,text))
    
    #Check for match
    if re.search(pattern,text):
        print('Match was found. \n')
    else:
        print('No Match was found.\n')

### Split with regular expressions
Let's see how we can split with the re syntax. This should look similar to how you used the split() method with strings.


In [0]:
import re

# Term to split on
split_term = '@'

phrase = 'What is the domain name of someone with the email: hello@gmail.com and bla@yahoo.com'

# Split the phrase
re.split(split_term,phrase)
#['What is the domain name of someone with the email: hello', 'gmail.com and bla','yahoo.com']


phrase.split('@')
#['What is the domain name of someone with the email: hello', 'gmail.com and bla','yahoo.com']


phrase.split('@',1)
#['What is the domain name of someone with the email: hello','gmail.com and bla@yahoo.com']



['What is the domain name of someone with the email: hello',
 'gmail.com and bla@yahoo.com']

In [0]:
phrase = 'What is the domain name of someone with the email: hello@gmail.com and bla@yahoo.com'

re.findall('@',phrase) #['@', '@']

re.search('@',phrase) #<_sre.SRE_Match object; span=(56, 57), match='@'>


<_sre.SRE_Match object; span=(56, 57), match='@'>

### re.sub(pattern, repl, string, count=0, flags=0)

Return the string obtained by replacing the leftmost non-overlapping occurrences of pattern in string by the replacement repl. If the pattern isn’t found, string is returned unchanged. repl can be a string or a function; if it is a string, any backslash escapes in it are processed. That is, \n is converted to a single newline character, \r is converted to a carriage return, and so forth. Unknown escapes of ASCII letters are reserved for future use and treated as errors. Other unknown escapes such as \& are left alone. Backreferences, such as \6, are replaced with the substring matched by group 6 in the pattern. 


### re.subn(pattern, repl, string, count=0, flags=0)
Perform the same operation as sub(), but return a tuple (new_string, number_of_subs_made).


## # re.findall()
Finding all instances of a pattern
You can use re.findall() to find all the instances of a pattern in a string. 
Pattern.findall(string[, pos[, endpos]])
Similar to the findall() function, using the compiled pattern, but also accepts optional pos and endpos parameters that limit the search region like for search().

In [0]:
input = 'hjgk fhfj abc dfd abcd   ukhkabc'

result = re.sub('abc',  '',    input)           # Delete pattern abc
print(result)
result = re.sub('abc',  'def', input)           # Replace pattern abc -> def
print(result)
result = re.sub(r'\s+', ' ',   input)           # Eliminate duplicate whitespaces
print(result)


hjgk fhfj  dfd d   ukhk
hjgk fhfj def dfd defd   ukhkdef
hjgk fhfj abc dfd abcd ukhkabc


In [0]:
:
def multi_re_find(patterns,phrase):
    '''
    Takes in a list of regex patterns
    Prints a list of all matches
    '''
    for pattern in patterns:
        print('Searching the phrase using the re check: %r' %(pattern))
        print(re.findall(pattern,phrase))
        print('\n')

#summary

**********************************
#### Repetition Syntax

```

test_phrase = 'sdsd..sssddd...sdddsddd...dsds...dsssss...sdddd'

test_patterns = [ 'sd*',     # s followed by zero or more d's
                'sd+',          # s followed by one or more d's
                'sd?',          # s followed by zero or one d's
                'sd{3}',        # s followed by three d's
                'sd{2,3}',      # s followed by two to three d's
                ]

```

### Character Sets



```
test_phrase = 'sdsd..sssddd...sdddsddd...dsds...dsssss...sdddd'

test_patterns = ['[sd]',    # either s or d
                's[sd]+']   # s followed by one or more s or d

multi_re_find(test_patterns,test_phrase)
```

### Exclusion


```
Use [^!.? ] to check for matches that are not a !,.,?, or space. Add a + to check that the match appears at least once. This basically translates into finding the words.


re.findall('[^!.? ]+',test_phrase)

['This',
 'is',
 'a',
 'string',
 'But',
 'it',
 'has',
 'punctuation',
 'How',
 'can',
 'we',
 'remove',
 'it']


```
### Character Ranges



```
test_phrase = 'This is an example sentence. Lets see if we can find some letters.'

test_patterns=['[a-z]+',      # sequences of lower case letters
               '[A-Z]+',      # sequences of upper case letters
               '[a-zA-Z]+',   # sequences of lower or upper case letters
               '[A-Z][a-z]+'] # one upper case letter followed by lower case letters
                
multi_re_find(test_patterns,test_phrase)
```

### Escape Codes



```

test_phrase = 'This is a string with some numbers 1233 and a symbol #hashtag'

test_patterns=[ r'\d+', # sequence of digits
                r'\D+', # sequence of non-digits
                r'\s+', # sequence of whitespace
                r'\S+', # sequence of non-whitespace
                r'\w+', # alphanumeric characters
                r'\W+', # non-alphanumeric
                ]

multi_re_find(test_patterns,test_phrase)
Searching the phrase using the re check: '\\d+'
['1233']
```




 


In [0]:
hex(12)

'0xc'

In [0]:
bin(128)

'0b10000000'

In [0]:
2**4

16

In [0]:
pow(2,3)

8

In [0]:
pow(2,4,3)

1

In [0]:
abs(-3)

3

In [0]:
round(3.9)

4

In [0]:
round(3.1)

3

In [0]:
round(3.14756,2)

3.15

In [0]:
import math
math.ceil(3.2)#4
a=math.pi+1
print(f"{a:0.2f}")

math.floor(3.9) #3
math.sqrt(4)#2.0

4.14


2.0

In [0]:
s="hello World!"
s.capitalize() #first word only in string
s.upper()
s.lower()
s.count('o') 
s.find('o') # gives us index

s.center(20,'#')
s.endswith('o')
s[-1] =='o'
s.split('i') #all instances are found aand removed (white space removal)
s.partition('i') #first instance only found and partitioned

In [0]:
s=set()
s.add(1)
s.add(2)
s.clear()
s={1,2,3}
sc=s.copy()
s.add(5)
s.discard(12)
s-sc or {1,2,3,5} - {1,2,3} #{5}


In [0]:
{1,2,3,5} - {1,2,3} #{5}

{5}

In [0]:
{1,2} | {5} #{1,2,5}

{1, 2, 5}

In [0]:
{1,2} & {1,4} #{1}

{1}

In [0]:
1,2,6 in {1,2,3,4}


(1, 2, False)

In [0]:
sc={1,2,3,4}
s={1,2}
s.issubset(sc)

True

''

In [0]:
{1,2} =={2,1}

True

In [0]:
{1,3} =={1,3}
{1,3} =={4,3}

False

In [0]:
s={1,2,3}
s.discard(12)
s.discard(1)
s

{2, 3}

In [0]:
d={'k1':1,'k2':2}
{x:x**2 for x in range(10)}

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81}

In [0]:
{k:v**2 for k,v in zip(['a','b'],range(2))}

{'a': 0, 'b': 1}

In [0]:
for k in d.items():
    print(k)

for k in d.keys():
    print(k)

for k in d.values():
    print(k)

('k1', 1)
('k2', 2)
k1
k2
1
2


In [0]:
l=[1,2,3]
l.append(5)
print(l)
l.append([4,5,6]) #append the whole itterable
print(l)
l.count(3) # whether or not it occures you won't get an error
print(l.count(10))
l.extend(['3']) #append the iterable one item at a time
print(l)
l.extend([4,5,6])
print(l)
l.index('3') #returns index of the first time it finds the item e.g '3' 
print(l)
'''
l.index('element not exist in list') #returns index of the first time it finds the item e.g '3' 
print('You get error')
'''
l.insert(0,'5') #['5', 1, 2, 3, 5, [4, 5, 6], '3', 4, 5, 6]
print(l)
l.pop()
print(l) #['5', 1, 2, 3, 5, [4, 5, 6], '3', 4, 5]
l.pop(0) # pops index 0 # [1, 2, 3, 5, [4, 5, 6], '3', 4, 5]
print(l)
l.remove(2) #  #inplace removes first instance of the item
print(l)
l.reverse() #inplace
print(l)
l=[5,76,8]
l.sort() #inplace
print(l)
sorted(reversed(l)) #neither of the reveresed or sorted are inplace
print(l)

[1, 2, 3, 5]
[1, 2, 3, 5, [4, 5, 6]]
0
[1, 2, 3, 5, [4, 5, 6], '3']
[1, 2, 3, 5, [4, 5, 6], '3', 4, 5, 6]
[1, 2, 3, 5, [4, 5, 6], '3', 4, 5, 6]
['5', 1, 2, 3, 5, [4, 5, 6], '3', 4, 5, 6]
['5', 1, 2, 3, 5, [4, 5, 6], '3', 4, 5]
[1, 2, 3, 5, [4, 5, 6], '3', 4, 5]
[1, 3, 5, [4, 5, 6], '3', 4, 5]
[5, 4, '3', [4, 5, 6], 5, 3, 1]
[5, 8, 76]
[5, 8, 76]


For more help you can watch tutorials here is link for discount 

[link text](https://docs.google.com/document/d/1TfB48iJcCAPaQYiOgB70_QgBXyGaU33nYyUZ6lgtLl8/edit)



In [0]:
' '.join(['1','2','3'])#'1 2 3'
''.join(['1','2','3'])#'123'
for i in enumerate([1,2,3]):    
    print(i)
# (0, 1)
# (1, 2)
# (2, 3)


(0, 1)
(1, 2)
(2, 3)
